api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,426 @@
+		}
+	}
+	// double check the sync status after 1 second to confirm (avoid false alarm)
+	time.Sleep(1 * time.Second)
+
+	otherHeight2 := ss.getMaxPeerHeight(false)
+	currentHeight := ss.blockChain.CurrentBlock().NumberU64()
+
+	isOutOfSync := currentHeight+inSyncThreshold < otherHeight2
+	utils.Logger().Info().
+		Uint64("OtherHeight1", otherHeight1).
+		Uint64("OtherHeight2", otherHeight2).
+		Uint64("lastHeight", lastHeight).
+		Uint64("currentHeight", currentHeight).
+		Msg("[SYNC] Checking sync status")
+	// Only confirm out of sync when the node has lower height and didn't move in heights for 2 consecutive checks
+	heightDiff := otherHeight2 - lastHeight
+	if otherHeight2 < lastHeight {
+		heightDiff = 0 // overflow
+	}
+	return SyncCheckResult{
+		IsInSync:    !(wasOutOfSync && isOutOfSync && lastHeight == currentHeight),
+		OtherHeight: otherHeight2,
+		HeightDiff:  heightDiff,
+	}
+}
+
+// GetActivePeerNumber returns the number of active peers
+func (ss *EpochSync) GetActivePeerNumber() int {
+	if ss.syncConfig == nil {
+		return 0
+	}
+	// len() is atomic; no need to hold mutex.
+	return len(ss.syncConfig.peers)
+}
+
+// SyncLoop will keep syncing with peers until catches up
+func (ss *EpochSync) SyncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, consensus *consensus.Consensus) {
+	timeout := 2
+	for {
+		<-time.After(time.Duration(timeout) * time.Second)
+		timeout = ss.syncLoop(bc, worker, isBeacon, consensus)
+	}
+}
+
+func (ss *EpochSync) syncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, _ *consensus.Consensus) (timeout int) {
+	maxHeight := ss.getMaxPeerHeight(isBeacon)
+	for {
+		block := bc.CurrentBlock()
+		height := block.NumberU64()
+		if height >= maxHeight {
+			utils.Logger().Info().
+				Msgf("[SYNC] Node is now IN SYNC! (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 10
+		}
+
+		utils.Logger().Info().
+			Msgf("[SYNC] Node is OUT OF SYNC (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+				isBeacon, bc.ShardID(), maxHeight, height)
+
+		var heights []uint64
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
+			}
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
+		}
+
+		if len(heights) == 0 {
+			// We currently on top.
+			return 10
+		}
+
+		err := ss.ProcessStateSync(heights, bc, worker)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 2
+		}
+	}
+}
+
+// getMaxPeerHeight gets the maximum blockchain heights from peers
+func (ss *EpochSync) getMaxPeerHeight(isBeacon bool) uint64 {
```

Comment:
Please refactor the code so no duplicated code like this exists.

Reply:
fixed